### ISSUE TITLE
predicted bpp is Nan when training

### ISSUE THREAD

**Issue by binzzheng** [InterDigitalInc/CompressAI#92]
## Bug

Hello, I built a video compressor with Compressai. Based on a simple hybrid coding framework, the video compressor uses hyper-prior entropy model to compress motion and residuals separately.  But when training, there will always be cases where the predicted bpp is _nan_ randomly. 

## Error

![error](https://user-images.githubusercontent.com/75297766/132681559-120667de-d8c2-4183-b682-2de1e8c1067c.png)

## Expected behavior

The hyper-prior entropy model will not predict bpp as _nan_ during training. 

## Environment

```
- PyTorch Version: 1.7.1
- CompressAI Version: 1.1.6
- OS: Ubuntu18.04
- Python version: 3.6.2
- CUDA/cuDNN version: 11.0 / 8005
```

## Additional context

I don't know why it will appear, and can't predict when it will appear. If I load a normal checkpoint and resume training again, it may not appear. If I finish the whole training process intermittently, the entropy model can also run normally.  Could you please provide me with some help?

**Comment by fracape** [comment]
Hi. The warning mentions the input tensor. Could it be corrupt input data with your training? Not easy to help without information on input motion, residuals and training data.

**Comment by binzzheng** [comment]
Thank you for your answer. I would try my best to describe some information, but I don't know if it is useful. 
#About training data
I used two consecutive frames in the video sequence as training data. It is simply processed before input, including normalization, random cropping, and random flipping. There should be no problem at this step. 
```python
def __getitem__(self, index):
        input_image = imageio.imread(self.image_input_list[index])
        ref_image = imageio.imread(self.image_ref_list[index])

        input_image = input_image.astype(np.float32) / 255.0
        ref_image = ref_image.astype(np.float32) / 255.0

        input_image = input_image.transpose(2, 0, 1)
        ref_image = ref_image.transpose(2, 0, 1)
        
        input_image = torch.from_numpy(input_image).float()
        ref_image = torch.from_numpy(ref_image).float()

        input_image, ref_image = random_crop_and_pad_image_and_labels(input_image, ref_image, [self.im_height, self.im_width])
        input_image, ref_image = random_flip(input_image, ref_image)

        return input_image, ref_image
```

#About input motion and residuals
The entropy model I use comes from _EntropyBottleneck_ and _GaussianConditional_ in Compressai. Other networks, such as _self.opticFlow_ and _sel.mvEncoder_, are CNNs with different structures. During training, I used _likelihoods_ from _EntropyBottleneck_ to calculate the predicted bpp. Combine the output _bpp_ and _mse_loss_ to train the entire video compression system. If any data is damaged, I think it may be in the process of calculating optical flow, such as _self.opticFlow_ or _self.motioncompensation_.
```python
def forward(self, input_image, referframe):
        estmv = self.opticFlow(input_image, referframe)
        mv_fea = self.mvEncoder(estmv)
        mv_prior = self.mvpriorEncoder(mv_fea)
        quant_mvprior, mvprior_likelihoods = self.entropy_hyper_mv(mv_prior)
        recon_mv_sigma = self.mvpriorDecoder(quant_mvprior)

        quant_mv = self.entropy_bottleneck_mv.quantize(mv_fea, "noise" if self.training else "dequantize")
        _, mv_likelihoods  = self.entropy_bottleneck_mv(mv_fea, recon_mv_sigma)
        recon_mv = self.mvDecoder(quant_mv)

        prediction, warpframe = self.motioncompensation(referframe, recon_mv)
        res = input_image - prediction
        res_fea = self.resEncoder(res)
        res_prior = self.respriorEncoder(res_fea)
        quant_resprior, resprior_likelihoods = self.entropy_hyper_res(res_prior)
        recon_res_sigma = self.respriorDecoder(quant_resprior)

        quant_res = self.entropy_bottleneck_res.quantize(res_fea, "noise" if self.training else "dequantize")
        _, res_likelihoods = self.entropy_bottleneck_res(res_fea, recon_res_sigma)

        recon_res = self.resDecoder(quant_res)
        recon_image = prediction + recon_res
        clipped_recon_image = recon_image.clamp(0. ,1.)

        mse_loss = torch.mean((recon_image - input_image).pow(2))
        warploss = torch.mean((warpframe - input_image).pow(2))
        interloss = torch.mean((prediction - input_image).pow(2))

        im_shape = input_image.size()
        batch_size = res_fea.size()[0]
        bpp_mv = torch.log(mv_likelihoods).sum() / (-math.log(2) * batch_size * im_shape[2] * im_shape[3])
        bpp_mvprior = torch.log(mvprior_likelihoods).sum() / (-math.log(2) * batch_size * im_shape[2] * im_shape[3])
        bpp_res = torch.log(res_likelihoods).sum() / (-math.log(2) * batch_size * im_shape[2] * im_shape[3])
        bpp_resprior = torch.log(resprior_likelihoods).sum() / (-math.log(2) * batch_size * im_shape[2] * im_shape[3])
        bpp = bpp_mv + bpp_mvprior + bpp_res + bpp_resprior
        
        return clipped_recon_image, mse_loss, warploss, interloss, bpp
```

#About training
I am not sure if I corrupted the original data during this training process. 
```python
for batch_idx, input in enumerate(train_loader):
        global_step += 1
        bat_cnt += 1
        input_image, ref_image = Var(input[0]), Var(input[1])
        clipped_recon_image, mse_loss, warploss, interloss, bpp = net(input_image, ref_image)

        mse_loss, warploss, interloss, bpp = torch.mean(mse_loss), torch.mean(warploss), torch.mean(interloss), torch.mean(bpp)
        distribution_loss = bpp
        distortion = mse_loss + warp_weight * (warploss + interloss)
        rd_loss = train_lambda * distortion + distribution_loss
        optimizer.zero_grad()
        aux_optimizer.zero_grad()
        rd_loss.backward()
        def clip_gradient(optimizer, grad_clip):
                for group in optimizer.param_groups:
                        for param in group["params"]:
                                if param.grad is not None:
                                        param.grad.data.clamp_(-grad_clip, grad_clip)
        clip_gradient(optimizer, 0.5)
        optimizer.step()
        
        aux_loss = net.aux_loss()
        aux_loss.backward()
        aux_optimizer.step()
```

**Comment by fracape** [comment]
Your last comment makes sense and is a good indication. I guess you can spot when the warnings happen first by printing more info and keeping the model as small and simple as possible. There seems to be 3 warnings at each problematic iteration.

**Comment by binzzheng** [comment]
Thanks for your advice! I will try to print more error information. If there are any new discoveries, I will update here. Thank you again!

**Comment by binzzheng** [comment]
**#Traning data**
The whole model was trained correctly for a complete epoch. So I think there should be no problem with the training data. 

**#When the warnings happen first**
When bpp appears as _nan_, I print out some variable values at this time. _estmv_ is completely correct. In _mv_fea_, part of it is _nan_ and part of it is the correct value. Obviously, the same situation also caused the _mv_prior_ part to be _nan_. So _nan_ may be caused in the process of passing through _self.mvEncoder_? 

‘’‘python
def forward(self, input_image, referframe):
        estmv = self.opticFlow(input_image, referframe)
        mv_fea = self.mvEncoder(estmv)
        mv_prior = self.mvpriorEncoder(mv_fea)
        quant_mvprior, mvprior_likelihoods = self.entropy_hyper_mv(mv_prior)
        recon_mv_sigma = self.mvpriorDecoder(quant_mvprior)

        quant_mv = self.entropy_bottleneck_mv.quantize(mv_fea, "noise" if self.training else "dequantize")
        _, mv_likelihoods  = self.entropy_bottleneck_mv(mv_fea, recon_mv_sigma)
        recon_mv = self.mvDecoder(quant_mv)

        prediction, warpframe = self.motioncompensation(referframe, recon_mv)
’‘’

_ResidualBlockWithStride_ and _ResidualBlock_ are from Compressai. The structure of my _self.mvEncoder_ is as follows: 

‘’‘python
class mvAnalysis(nn.Module):
        def __init__(self):
                super(mvAnalysis, self).__init__()
                self.RB1 = ResidualBlockWithStride(2, out_channel, stride=2)
                self.RB2 = ResidualBlock(out_channel, out_channel)
                self.RB3 = ResidualBlockWithStride(out_channel, out_channel, stride=2)
                self.RB4 = ResidualBlock(out_channel, out_channel)
                self.RB5 = ResidualBlockWithStride(out_channel, out_channel, stride=2)
                self.conv = conv3x3(out_channel, out_channel, stride=2)

        def forward(self, x):
                x = self.RB1(x)
                x = self.RB2(x)
                x = self.RB3(x)
                x = self.RB4(x)
                x = self.RB5(x)
                out = self.conv(x)
                return out
’‘’

Some printed numerical information, from top to bottom, are _estmv_, _mv_fea_ and _mv_prior_. When error occurs, _estmv_ is correct, but _mv_fea_ and _mvprior_ are wrong. 
![estmv](https://user-images.githubusercontent.com/75297766/133051067-6b9ad1dd-105c-4018-9078-eec6ae148a30.png)
![mv_fea](https://user-images.githubusercontent.com/75297766/133051082-1735c625-731e-481b-a11c-24592f2f1196.png)
![mv_prior](https://user-images.githubusercontent.com/75297766/133051100-2dabe7f9-d482-40e0-b97e-471959c49f5c.png)

**#Additional information**
During training, _rd_loss_ and _aux_loss_ are iterated with two optimizers. Usually, I set the learning rate of _optimizer_ (for _rd_loss_) and _aux_optimizer_ (for _aux_loss_) to 0.0001 and 0.001, respectively. While keeping the learning rate of _aux_optimizer_ at 0.001, I set the learning rate of _optimizer_ to 0. Even so, _nan_ still happened. I don’t know if my training settings are wrong. 

‘’‘python
        distribution_loss = bpp
        distortion = mse_loss + warp_weight * (warploss + interloss)
        rd_loss = train_lambda * distortion + distribution_loss
        optimizer.zero_grad()
        aux_optimizer.zero_grad()
        rd_loss.backward()
        def clip_gradient(optimizer, grad_clip):
                for group in optimizer.param_groups:
                        for param in group["params"]:
                                if param.grad is not None:
                                        param.grad.data.clamp_(-grad_clip, grad_clip)
        clip_gradient(optimizer, 0.5)
        optimizer.step()
        
        aux_loss = net.aux_loss()
        aux_loss.backward()
        aux_optimizer.step()
‘’‘

**Comment by fracape** [comment]
Not sure it's related, but have you tried disabling clip_gradient? Or use torch.nn.utils.clip_grad_norm_ for instance?

**Comment by fracape** [comment]
ok thanks for the feedback. Going to close this, since it relates to a side use case and does not break for image compression. Feel free to post in the section discussions to get additional help from other users.
